src/pino_ode_solve.jl

```diff
@@ -0,0 +1,217 @@
+* Sifan Wang "Learning the solution operator of parametric partial differential equations with physics-informed DeepOnets"
+* Zongyi Li "Physics-Informed Neural Operator for Learning Partial Differential Equations"
+"""
+struct PINOODE{C, O, B, I, S <: Union{Nothing, AbstractTrainingStrategy},
+    AL <: Union{Nothing, Function}, K} <:
+       SciMLBase.AbstractODEAlgorithm
+    chain::C
+    opt::O
+    bounds::B
+    init_params::I
+    strategy::S
+    additional_loss::AL
+    kwargs::K
+end
+
+function PINOODE(chain,
+        opt,
+        bounds;
+        init_params = nothing,
+        strategy = nothing,
+        additional_loss = nothing,
+        kwargs...)
+    !(chain isa Lux.AbstractExplicitLayer) && (chain = Lux.transform(chain))
+    PINOODE(chain, opt, bounds, init_params, strategy, additional_loss, kwargs)
+end
+mutable struct PINOPhi{C, S}
+    chain::C
+    st::S
+    function PINOPhi(chain::Lux.AbstractExplicitLayer, st)
+        new{typeof(chain), typeof(st)}(chain, st)
+    end
+end
+
+function generate_pino_phi_θ(chain::Lux.AbstractExplicitLayer, init_params)
+    θ, st = Lux.setup(Random.default_rng(), chain)
+    if init_params === nothing
+        init_params = ComponentArrays.ComponentArray(θ)
+    else
+        init_params = ComponentArrays.ComponentArray(init_params)
+    end
+    PINOPhi(chain, st), init_params
+end
+
+function (f::PINOPhi{C, T})(x::NamedTuple, θ) where {C <: NeuralOperator, T}
+    y, st = f.chain(adapt(parameterless_type(ComponentArrays.getdata(θ)), x), θ, f.st)
+    ChainRulesCore.@ignore_derivatives f.st = st
+    y
+end
+
+function dfdx(phi::PINOPhi{C, T}, x::Tuple, θ, prob::ODEProblem) where {C <: DeepONet, T}
+    p, t = x
+    f = prob.f
+    branch_left, branch_right = f.(0, p, t .+ sqrt(eps(eltype(p)))), f.(0, p, t)
+    trunk_left, trunk_right = t .+ sqrt(eps(eltype(t))), t
+    x_left = (branch = branch_left, trunk = trunk_left)
+    x_right = (branch = branch_right, trunk = trunk_right)
+    (phi(x_left, θ) .- phi(x_right, θ)) / sqrt(eps(eltype(t)))
+end
+
+function physics_loss(
+        phi::PINOPhi{C, T}, prob::ODEProblem, x, θ) where {C <: DeepONet, T}
+    p, t = x
+    f = prob.f
+    #TODO If du = f(u,p,t), where f = g(p,t)*u so it will wrong, f(0, p, t) = g(p,t)*0 = 0
+    #work correct only with function like du = f(p,t) + g(u)
+    du = vec(dfdx(phi, x, θ, prob))
+
+    tuple = (branch = f.(0, p, t), trunk = t)
+    out = phi(tuple, θ)
+    f_ = vec(f.(out, p, t))
+    norm = prod(size(out))
+    sum(abs2, du .- f_) / norm
+end
+
+function operator_loss(
+        phi::PINOPhi{C, T}, prob::ODEProblem, x, θ) where {C <: DeepONet, T}
+    p, t = x
+    f = prob.f
+    t0 = t[:, :, [1]]
+    f_0 = f.(0, p, t0)
+    tuple = (branch = f_0, trunk = t0)
+    out = phi(tuple, θ)
+    u = vec(out)
+    u0_ = fill(prob.u0, size(out))
+    u0 = vec(u0_)
+    norm = prod(size(u0_))
+    sum(abs2, u .- u0) / norm
+end
+
+function get_trainset(strategy::GridTraining, bounds, tspan)
```

Comment:
Just reuse the non pino?

Reply:
here, function that is create a set of training data is for both the physics loss function and the loss function for the initial condition ||u(0)-u0|| (it was named operator_loss I renamed it to inital_condition_loss)